### PR TITLE
[infra-proxy-service] Manage node tags with add, delete and set actions

### PR DESCRIPTION
**Pull request by vsingh-msys**
### :nut_and_bolt: Description: What code changed, and why?

In order to manage the tags, we have three sets of action as follows:

 - `add`: Adds the tags to existing node tags.
 - `delete`: Deletes the tags present on the node.
 - `set`: Reset the tags with given tags for the node.

Also did some code-refactoring for `UpdateNode` and `DeleteNode` proto request format. These APIs under development so no impact on proto changes. [02/22/2020 Updated]

### :chains: Related Resources
Fixes: #4530 

### :+1: Definition of Done

- Infra service is able to manage the node tags using add, delete, and set operations.

### :athletic_shoe: How to Build and Test the Change

Steps to build:
 - `rebuild components/infra-proxy-service`
 - `rebuild components/automate-gateway`

Steps to test:

- To add data using 3rd internal steps https://github.com/chef/automate/blob/master/dev-docs/adding-data/adding_test_data.md#adding-data-to-infra-views

- Node tags add API:

```
 curl -sSkH "api-token: $(get_admin_token)" -X PUT https://a2-dev.test/api/v0/infra/servers/chef-server-dev-test/orgs/chef-org-ev/nodes/aix-7.1/tags?pretty  -d '{"action": "add", "tags": ["tag1"] }'
{
  "tags": [
    "aix7.1",
    "aix",
    "ibm",
    "tag1"
  ]
}
```

- Node tags remove API:
```
curl -sSkH "api-token: $(get_admin_token)" -X PUT https://a2-dev.test/api/v0/infra/servers/chef-server-dev-test/orgs/chef-org-ev/nodes/aix-7.1/tags?pretty  -d '{"action": "delete", "tags": ["tag1"] }'
{
  "tags": [
    "aix7.1",
    "aix",
    "ibm"
  ]
}
```

- Node tags set API:
```
 curl -sSkH "api-token: $(get_admin_token)" -X PUT https://a2-dev.test/api/v0/infra/servers/chef-server-dev-test/orgs/chef-org-ev/nodes/aix-7.1/tags?pretty  -d '{"action": "set", "tags": ["ibm"] }'
{
  "tags": [
    "ibm"
  ]
}
```

### :white_check_mark: Checklist

- [x] I have read the [CONTRIBUTING document](https://github.com/chef/automate/blob/master/CONTRIBUTING.md).
- [x] Tests added/updated?
- [ ] Docs added/updated?
- [x] All commits have been signed-off for the [Developer Certification of Origin](https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco).

### :camera: Screenshots, if applicable
Signed-off-by: Vivek Singh <vivek.singh@msystechnologies.com>

### CODE REVIEW

File: components/infra-proxy-service/server/nodes.go [hunk 1]
```diff
@@ -191,6 +191,56 @@ func (s *Server) UpdateNode(ctx context.Context, req *request.UpdateNode) (*resp
...
+		return nil, ParseAPIError(err)
+	}
+
+	tags := SafeSliceFromMap(chefNode.NormalAttributes, "tags")
+	switch req.Action {
+	case "add":
+		tags = Unique(append(tags, req.Tags...))
+	case "delete":
+		tags = SubstractSlice(tags, req.Tags)
+	case "set":
```
Comment (punitmundra) [inline components/infra-proxy-service/server/nodes.go:227]: does add will do append the remaining list and set will create the new set of tag ?

**Review by punitmundra — APPROVED** [review]

File: components/infra-proxy-service/server/nodes.go (hunk 1 above)
Comment (vsingh-msys) [reply components/infra-proxy-service/server/nodes.go:227]: Yes, add will be used for appending to existing tags, set will reset/create a new set of tags.

**Review by shaik80 — APPROVED** [review]

**Review by shaik80 — CHANGES_REQUESTED** [review]

File: api/external/infra_proxy/infra_proxy.proto [hunk 2]
```diff
@@ -255,12 +255,12 @@ service InfraProxy {
...
     option (chef.automate.api.iam.policy).action = "infra:infraServers:get";
   };
-  rpc DeleteNode (infra_proxy.request.DeleteNode) returns (infra_proxy.response.DeleteNode) {
+  rpc DeleteNode (infra_proxy.request.Node) returns (infra_proxy.response.DeleteNode) {
     option (google.api.http).delete = "/api/v0/infra/servers/{server_id}/orgs/{org_id}/nodes/{name}";
     option (chef.automate.api.iam.policy).resource = "infra:infraServers:{server_id}:orgs:{org_id}:nodes";
     option (chef.automate.api.iam.policy).action = "infra:infraServers:update";
   };
-  rpc UpdateNode (infra_proxy.request.UpdateNode) returns (infra_proxy.response.UpdateNode) {
+  rpc UpdateNode (infra_proxy.request.NodeObject) returns (infra_proxy.response.Node) {
```
Comment (vsingh-msys) [inline api/external/infra_proxy/infra_proxy.proto:263]: `UpdateNode` & `CreateNode` now using same request `NodeObject` so we can reuse it.

File: api/external/infra_proxy/request/nodes.proto [hunk 3]
```diff
@@ -34,16 +34,7 @@ message AffectedNodes {
...
 }
 
-message DeleteNode {
-    // Chef organization ID.
-    string org_id = 1;
-    // Chef Infra Server ID.
-    string server_id = 2;
-    // Node name.
-    string name = 3;
-}
```
Comment (vsingh-msys) [inline api/external/infra_proxy/request/nodes.proto:44]: Use `Node` in place of `DeleteNode` api.

File: components/infra-proxy-service/server/nodes.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (vsingh-msys) [reply components/infra-proxy-service/server/nodes.go:208]: Added required integration tests https://github.com/chef/automate/blob/VSingh/infra-proxy-service-manage-node-tags/components/infra-proxy-service/integration_test/nodes_test.go

**Comment by vsingh-msys** [comment]
- Added `nodes_test.go` file to cover the integration test cases. Also added interservice `CreateNode` for adding the test cases.


This PR is ready for re-review please have look.

File: components/infra-proxy-service/integration_test/nodes_test.go [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+		// Add tag3 into existing the tags
+		req.Tags = []string{"tag3"}
+		req.Action = "add"
+		res, err = infraProxy.UpdateNodeTags(ctx, req)
+		assert.NoError(t, err)
+		assert.NotNil(t, res)
+		assert.Equal(t, []string{"tag1", "tag2", "tag3"}, res.Tags)
+	})
+
+	t.Run("when set tags request with valid params", func(t *testing.T) {
```
Comment (msorens) [inline components/infra-proxy-service/integration_test/nodes_test.go:87]:
```suggestion
	t.Run("when set, replaces tags with specified values", func(t *testing.T) {
```

File: components/infra-proxy-service/integration_test/nodes_test.go [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+
+		res, err := infraProxy.GetNode(ctx, req)
+		assert.Error(t, err)
+		assert.Nil(t, res)
+	})
+}
+
+func TestUpdateNodeTags(t *testing.T) {
+	ctx := context.Background()
+	t.Run("when add tags request with valid params", func(t *testing.T) {
```
Comment (msorens) [inline components/infra-proxy-service/integration_test/nodes_test.go:54]:
```suggestion
	t.Run("when add, adds values to the current group of tags", func(t *testing.T) {
```

File: components/infra-proxy-service/integration_test/nodes_test.go [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+		// Set the tags with the tag reset1
+		req.Tags = []string{"reset1"}
+		req.Action = "set"
+		res, err = infraProxy.UpdateNodeTags(ctx, req)
+		assert.NoError(t, err)
+		assert.NotNil(t, res)
+		assert.Equal(t, []string{"reset1"}, res.Tags)
+	})
+
+	t.Run("when delete tags request with valid params", func(t *testing.T) {
```
Comment (msorens) [inline components/infra-proxy-service/integration_test/nodes_test.go:120]:
```suggestion
	t.Run("when delete, removes tags from the current group of values", func(t *testing.T) {
```

File: components/infra-proxy-service/integration_test/nodes_test.go [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+			Name:     fmt.Sprintf("no-node-%d", time.Now().Nanosecond()),
+		}
+
+		res, err := infraProxy.GetNode(ctx, req)
+		assert.Error(t, err)
+		assert.Nil(t, res)
+	})
+}
+
+func TestUpdateNodeTags(t *testing.T) {
```
Comment (msorens) [inline components/infra-proxy-service/integration_test/nodes_test.go:52]: Additional tests suggested:
 * when attempting to delete a non-existent tag, leaves tag group unchanged (?? or throws an error, whatever is correct here)
 * when attempting to delete all tags, updates tag group to the empty list
 * when attempting to delete some existing and some non-existing tags simultaneously, does ??
 * when attempting to set tags to empty list, updates tag group to the empty list

File: components/infra-proxy-service/server/util.go [hunk 8]
```diff
@@ -39,3 +39,38 @@ func SafeSliceFromMap(values map[string]interface{}, key string) []string {
 
 	return []string{}
 }
+
+// SubstractSlice return the slice without containing all element of b.
+func SubstractSlice(a []string, b []string) []string {
```
Comment (msorens) [inline components/infra-proxy-service/server/util.go:44]: Fix typo and make the explanation more clear.
```suggestion
// SubtractSlice returns the slice a with all elements of b removed.
func SubtractSlice(a []string, b []string) []string {
```

File: components/infra-proxy-service/server/util_test.go [hunk 9]
```diff
@@ -44,3 +44,28 @@ func TestSafeSliceFromMap(t *testing.T) {
 	assert.Equal(t, []string{}, SafeSliceFromMap(testData, "test3"))
 	assert.Equal(t, []string{}, SafeSliceFromMap(testData, "no-key"))
 }
+
+func TestSubstractSlice(t *testing.T) {
```
Comment (msorens) [inline components/infra-proxy-service/server/util_test.go:48]: Add tests:
* subtract both tag3 and unknown from the list of tag1, tag2, and tag3.
* subtract empty list from a non-empty list
* subtract a non-empty list from an empty list
* subtract whole list from itself

File: components/infra-proxy-service/server/util_test.go [hunk 10]
```diff
@@ -44,3 +44,28 @@ func TestSafeSliceFromMap(t *testing.T) {
...
+	assert.Equal(t, []string{"tag1", "tag2"}, RemoveElement(testData1, "tag3"))
+
+	testData2 := []string{"tag1", "tag2", "tag3"}
+	assert.Equal(t, testData2, RemoveElement(testData2, "unknown"))
+
+	testData3 := []string{"tag1", "tag2", "tag3"}
+	assert.Equal(t, testData3, RemoveElement(testData3, ""))
+}
+
+func TestUnique(t *testing.T) {
```
Comment (msorens) [inline components/infra-proxy-service/server/util_test.go:65]: Add tests:
 * tag1, tag2, tag3, tag1 (i.e. non-adjacent in the list)
 * tag, tag, tag (i.e. all elements are the same)
 * empty list

File: components/infra-proxy-service/server/util_test.go [hunk 11]
```diff
@@ -44,3 +44,28 @@ func TestSafeSliceFromMap(t *testing.T) {
...
 	assert.Equal(t, []string{}, SafeSliceFromMap(testData, "no-key"))
 }
+
+func TestSubstractSlice(t *testing.T) {
+	assert.Equal(t, []string{"tag2", "tag3"}, SubstractSlice([]string{"tag1", "tag2", "tag3"}, []string{"tag1"}))
+	assert.Equal(t, []string{"tag3"}, SubstractSlice([]string{"tag1", "tag2", "tag3"}, []string{"tag1", "tag2"}))
+	assert.Equal(t, []string{"tag1", "tag2", "tag3"}, SubstractSlice([]string{"tag1", "tag2", "tag3"}, []string{"unknown"}))
+}
+
+func TestRemoveElement(t *testing.T) {
```
Comment (msorens) [inline components/infra-proxy-service/server/util_test.go:54]: Add tests:
* Remove value from empty list
* Remove empty string from empty list

File: components/infra-proxy-service/integration_test/nodes_test.go [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+
+	t.Run("when the node does not exist, return node not found", func(t *testing.T) {
+		req := &request.Node{
+			ServerId: autoDeployedChefServerID,
+			OrgId:    autoDeployedChefOrganizationID,
+			Name:     fmt.Sprintf("no-node-%d", time.Now().Nanosecond()),
+		}
+
+		res, err := infraProxy.GetNode(ctx, req)
+		assert.Error(t, err)
```
[the remark this answers is not in the document]
Comment (vsingh-msys) [reply components/infra-proxy-service/integration_test/nodes_test.go:47]: oh yeah, forgot to add this, thanks for pointing it out.

File: components/infra-proxy-service/server/util_test.go (hunk 9 above)
Comment (vsingh-msys) [reply components/infra-proxy-service/server/util_test.go:48]: Thank you for providing much detailed test cases scenario!

**Comment by vsingh-msys** [comment]
@msorens Thank you for reviewing PR in detail, Incorporated the reviews comments and updated the PR.  

Some of the reviews could be minimized from my side, will take this into consideration for further PRs Thanks!

**Review by msorens — APPROVED** [review]
Nice work, @vsingh-msys !

**Pull request merged**